% Generated by roxygen2 (4.0.2): do not edit by hand
\name{bedgraph2norm}
\alias{bedgraph2norm}
\title{Import bedgraph to GRanges}
\usage{
bedgraph2norm(bedgraph_file, fasta_file, txDb, bed_file,
  column_name = "bedgraph_score", add_to)
}
\arguments{
\item{bedgraph_file}{path to compatible bedgraph file}

\item{fasta_file}{path to fasta file which is used for a) choosing which
transcripts to use (transcripts absent from fasta are not reported), b)
providing sequence for to display in GRanges metadata}

\item{txDb}{TranscriptDb object with transcript definitions. Names must
match those in fasta_file}

\item{bed_file}{character containing file path to BED file with transcript
definitions. Supply txDb XOR bedfile}

\item{column_name}{How to name imported metadata in GRanges}

\item{add_to}{GRanges object made by other normalization function (dtcr(),
slograt(), swinsor(), compdata()) to which values from bedgraph should be
added.}
}
\value{
Function creates GRanges object or (if add_to specified) adds
metadata to already existing object
}
\description{
Function importing data from bedgraph format compatible with UCSC Genome
Browser to norm_GR data frame. Warning: Compatible only with bedgraph files
generated by norm2bedgraph function - bedgraph need to have 2 tracks, first
for plus strand, second for minus strand.
}
\details{
%% ~~ If necessary, more details than the description above ~~
}
\note{
%% ~~further notes~~
}
\examples{
##---- Should be DIRECTLY executable !! ----
##-- ==>  Define data, use random,
##-- or do  help(data=index)  for the standard data sets.
}
\author{
Lukasz Jan Kielpinski
}
\references{
%% ~put references to the literature/web site here ~
}
\seealso{
\code{\link{norm2bedgraph}}, \code{\link{GR2norm_df}}, \code{\link{plotRNA}},
\code{\link{BED2txDb}}, \code{\link{dtcr}}, \code{\link{slograt}}, \code{\link{swinsor}},
\code{\link{compdata}}
}
\keyword{~kwd1}
\keyword{~kwd2}

