% Generated by roxygen2 (4.0.2): do not edit by hand
\name{plotReads}
\alias{plotReads}
\title{Function plotting ranges from GRanges}
\usage{
plotReads(euc_GR, RNAid, cutoff = 1, order_by = 1, ylab, xlab, main, ylim,
  xlim, ...)
}
\arguments{
\item{euc_GR}{GRanges generated by readsamples() function}

\item{RNAid}{Transcript identifier - for which transcript plot should be
generated.}

\item{cutoff}{specifies cutoff length, only inserts of this length or longer
will be used for processing (default: 1)}

\item{order_by}{how displayed reads in plotReads function should be sorted.
1 - for sorting by termination location, 2 for sorting by reverse
transcription start site}
}
\value{
Plotting function.
}
\description{
Function plots spans of cDNA inserts from GRanges created by readsamples()
function. Similar to Figure 4A in reference 1.
}
\details{
%% ~~ If necessary, more details than the description above ~~
}
\note{
%% ~~further notes~~
}
\examples{
dummy_euc_GR <- GRanges(seqnames="DummyRNA", IRanges(start=round(runif(100)*100), width=round(runif(100)*100+1)), strand="+", EUC=round(runif(100)*100))
plotReads(dummy_euc_GR, RNAid="DummyRNA")
}
\author{
Lukasz Jan Kielpinski
}
\references{
Kielpinski, L.J., and Vinther, J. (2014). Massive
parallel-sequencing-based hydroxyl radical probing of RNA accessibility.
Nucleic Acids Res.
}
\seealso{
\code{\link{plot}}, \code{\link{plot.default}}, \code{\link{readsamples}}
}
\keyword{~kwd1}
\keyword{~kwd2}

