###Function importing data from bedgraph format compatible with UCSC Genome Browser to norm_GR data frame.
# bedgraph file requirements: it need to contain 2 tracks, first - scores for plus strand, second - scores for minus strand
# Description of options:
# bedgraph_file - file path to compatible bedgraph file
# fasta_file - file path to fasta file which is used for a) choosing which transcripts to use (transcripts absent from fasta are not reported), b) provide sequences for norm_df
# txDb - TranscriptDb object with transcript definitions. Names must match those in fasta_file
# bed_file - character containing file path to BED file with transcript definitions. Supply txDb XOR bedfile
# column_name - name of the column in norm_GR containing scores from bedgraph
# add_to - normalized data frame with already performed normalization of another kind. Results will be merged
# Function creates norm_GR GRanges object
# Warning: Compatible only with bedgraph files generated by norm2bedgraph function - bedgraph need to have 2 tracks, first for plus strand, second for minus strand.













#' Import bedgraph to GRanges
#'
#' Function importing data from bedgraph format compatible with UCSC Genome
#' Browser to norm_GR data frame. Warning: Compatible only with bedgraph files
#' generated by norm2bedgraph function (bedgraph needs to have 2 tracks, first
#' for plus strand, second for minus strand). May be used for transforming
#' normalized data to another different annotation sets.
#'
#' %% ~~ If necessary, more details than the description above ~~
#'
#' @param bedgraph_file path to compatible bedgraph file
#' @param fasta_file path to fasta file which is used for a) choosing which
#' transcripts to use (transcripts absent from fasta are not reported), b)
#' providing sequence for to display in GRanges metadata
#' @param txDb TranscriptDb object with transcript definitions. Names must
#' match those in fasta_file
#' @param bed_file character containing file path to BED file with transcript
#' definitions. Supply txDb XOR bedfile
#' @param column_name How to name imported metadata in GRanges
#' @param add_to GRanges object made by other normalization function (dtcr(),
#' slograt(), swinsor(), compdata()) to which values from bedgraph should be
#' added.
#' @return Function creates GRanges object or (if add_to specified) adds
#' metadata to already existing object
#' @note %% ~~further notes~~
#' @author Lukasz Jan Kielpinski
#' @seealso \code{\link{norm2bedgraph}}, \code{\link{GR2norm_df}}, \code{\link{plotRNA}},
#' \code{\link{BED2txDb}}, \code{\link{dtcr}}, \code{\link{slograt}}, \code{\link{swinsor}},
#' \code{\link{compdata}}
#' @references %% ~put references to the literature/web site here ~
#' @keywords ~kwd1 ~kwd2
#' @examples
#'
#' ##---- Should be DIRECTLY executable !! ----
#' ##-- ==>  Define data, use random,
#' ##-- or do  help(data=index)  for the standard data sets.
#'
#' @import rtracklayer Biostrings
#' @export bedgraph2norm
bedgraph2norm <- function(bedgraph_file, fasta_file, txDb,  bed_file, column_name="bedgraph_score", add_to){

###Check conditions:
if(missing(txDb) & missing(bed_file))
    stop("Error: specify gene annotation")

if(missing(fasta_file))
    stop("Specify fasta file")

if(!file.exists(fasta_file))
    stop("Fasta file not found")

###Define functions:
    #Function to import the sequence from fasta file, if no sequence - fill in with N's:
    add_sequence <- function(oneRNA_norm){
        if(max(oneRNA_norm$Pos) > length(txs[[as.character(oneRNA_norm$RNAid[1])]])){
            unexpected_length_difference <- max(oneRNA_norm$Pos) - length(txs[[as.character(oneRNA_norm$RNAid[1])]])
            one_gene_sequence <- unlist(strsplit(as.character(c(txs[[as.character(oneRNA_norm$RNAid[1])]], DNAString(paste(rep("N",unexpected_length_difference), collapse="")))[oneRNA_norm$Pos]), ""))
            message(paste("For RNA ", oneRNA_euc[1,1], "positions outside FASTA annotation exist. N's added"))
        }else{
            one_gene_sequence <- unlist(strsplit(as.character(txs[[as.character(oneRNA_norm$RNAid[1])]][oneRNA_norm$Pos]), ""))
            }
        oneRNA_norm$nt <- one_gene_sequence
        oneRNA_norm
        }
###

#Read in bedgraph file and convert both strands to single UCSCdata entry:
bedgraph_list <- import.bedGraph(bedgraph_file)
if(length(bedgraph_list)!=2){stop("Error: bedgraph file needs to contain 2 tracks, one for each strand.")}
strand(bedgraph_list[[1]]) <- "+"
strand(bedgraph_list[[2]]) <- "-"
bedgraph_merged <- c(bedgraph_list[[1]], bedgraph_list[[2]])

##Prepare GRangesList for mapping values to:
txs <- readDNAStringSet(fasta_file) #read in fasta file.
if(missing(txDb)){txDb <- BED2txDb(bed_file)} #if txDb not given, make it from bed file
all_exons <- exonsBy(txDb, "tx", use.names=TRUE) #build GRangesList from txDb
my_exons <- all_exons[names(all_exons) %in% names(txs)] #keep in GRangesList only those transcript that are present in fasta file

#Print warning if transcripts overlap:
overlapping_transcripts <- which(countOverlaps(my_exons) > 1)
if(length(overlapping_transcripts) > 0){
    message(paste("Warning: transcript",names(my_exons)[overlapping_transcripts], "overlaps with another transcript. Score is added to more than one RNA."))
    }

#Map bedgraph positions to annotated transcripts and format to norm_df:
mapped_to_tx <- map(bedgraph_merged, my_exons)
hits_in_tx <- subjectHits(mapped_to_tx) #Transcripts with signal
hits_in_EF <- queryHits(mapped_to_tx) #Which signal given transcript has
normalized <- data.frame(RNAid=names(my_exons)[hits_in_tx],Pos=start(ranges(mapped_to_tx)), nt=NA, bedgraph_score=score(bedgraph_merged)[hits_in_EF]) #Data frame with results
names(normalized)[names(normalized)=="bedgraph_score"] <- column_name #change column name to one provided by user

###Add sequence information from fasta file (on per RNA basis)
norm_by_RNA <- split(normalized, f=normalized$RNAid, drop=TRUE)
normalized <- do.call(rbind, lapply(norm_by_RNA, FUN=add_sequence))

#If add_to specified, merge with existing normalized data frame:
if(!missing(add_to)){
add_to_df <- GR2norm_df(add_to)
normalized <- merge(add_to_df, normalized, by=c("RNAid", "Pos", "nt"), suffixes=c(".old",".new"))
}
###

normalized <- normalized[order(normalized$RNAid, normalized$Pos),]
normalized_GR <- norm_df2GR(normalized)

normalized_GR
}

