###Function in the normalizing functions family.
# Performs sliding window Winsorization given treated Comp_GR generated by comp() function. It winsorizes values in windows (of a size specified by window_size) sliding by 1 nt over whole transcript length and reports average winsorized value for each nucleotide.
# Description of options:
# winsor_level - where should winsorization boundaries be made. Value 0.9 indicates that top 5% of data will be reduced to 95% quantile and bottom 5% to 5% quantile (default: 0.9)
# window_size - window size for Winsorazation (default: 71)
# nt_offset - how many nucleotides before modification the reverse transcription terminates (default: 0)
# only_top - if TRUE then bottom quantile is fixed to 0 (default= FALSE)
# add_to - normalized GRanges with already performed normalization of another kind. Results will be merged












#' Smooth Winsorization
#'
#' Performs sliding window Winsorization given treated GRanges generated by
#' comp() function. It winsorizes values in windows (of a size specified by
#' window_size) sliding by 1 nt over whole transcript length and reports mean
#' winsorized value for each nucleotide (as well as standard deviation).
#'
#' %% ~~ If necessary, more details than the description above ~~
#'
#' @param Comp_GR GRanges object made by comp() function.
#' @param winsor_level Winsorization level. Bottom outliers will be set to
#' (1-winsor_level)/2 quantile and top outliers to (1+winsor_level)/2 quantile.
#' @param window_size Size of a sliding window.
#' @param only_top If TRUE then bottom values are not Winsorized and are set to
#' 0.
#' @param nt_offset How many position in the 5' direction should the signal be
#' offset to account for the fact that reverse transcription termination occurs
#' before site of modification.
#' @param add_to GRanges object made by other normalization function (dtcr(),
#' slograt(), swinsor(), compdata()) to which normalized values should be
#' added.
#' @return GRanges object with "swinsor" (mean smooth-Winsor values) and
#' "swinsor.sd" (standard deviation of smooth-Winsor values) metadata.
#' @note %% ~~further notes~~
#' @author Lukasz Jan Kielpinski
#' @seealso \code{\link{comp}}, \code{\link{dtcr}}, \code{\link{slograt}},
#' \code{\link{compdata}}, \code{\link{GR2norm_df}},
#' \code{\link{plotRNA}}, \code{\link{norm2bedgraph}}
#' @references SHAPES publication. Poulsen, Kielpinski, Salama, Krogh, Vinther.
#' In review as for 2nd Oct 2014.
#' @keywords ~kwd1 ~kwd2
#' @examples
#'
#' dummy_euc_GR <- GRanges(seqnames="DummyRNA", IRanges(start=round(runif(100)*100), width=round(runif(100)*100+1)), strand="+", EUC=round(runif(100)*100))
#' dummy_comp_GR <- comp(dummy_euc_GR)
#' swinsor(dummy_comp_GR)
#'
#' @import GenomicRanges
#' @export swinsor
swinsor <- function(Comp_GR, winsor_level=0.9, window_size=71, only_top= FALSE, nt_offset=1, add_to){

###Check conditions:
    if(nt_offset < 0){
    print("error: nt_offset must be >= 0")
    stop()
    }
###Define functions:
#Process single RNA:
swinsor_oneRNA <- function(oneRNA_comp_df){
    if(prod(diff(oneRNA_comp_df$Pos)==rep(1, nrow(oneRNA_comp_df)-1))==1){ #Checks if data is properly sorted.
        means_and_sds <- swinsor_vector(oneRNA_comp_df$TC, window_size=window_size, winsor_level=winsor_level, only_top=only_top)
        oneRNA_comp_df$swinsor <- means_and_sds[[1]][(1+nt_offset):(nrow(oneRNA_comp_df)+nt_offset)]
        oneRNA_comp_df$swinsor.sd <- means_and_sds[[2]][(1+nt_offset):(nrow(oneRNA_comp_df)+nt_offset)]
        return(oneRNA_comp_df[1:(nrow(oneRNA_comp_df)-nt_offset),])}else{
    print(paste("Check if data was properly sorted by comp() function. Problem with",oneRNA_comp_df$RNAid[1]))
    stop()
    }
    }

    ###Function body:
    Comp_df <- GR2norm_df(Comp_GR)

    Comp_df[is.na(Comp_df)] <- 0 #Changes NA to 0 - treat no events as 0 events.

    Comp_df <- Comp_df[order(Comp_df$RNAid, Comp_df$Pos),]
    Comp_df_by_RNA <- split(Comp_df, f=Comp_df$RNAid, drop=TRUE)

    normalized <- do.call(rbind, lapply(Comp_df_by_RNA, FUN=swinsor_oneRNA)) #Do processing (smooth, offset, p-values)
    normalized <- data.frame(RNAid=normalized$RNAid, Pos=normalized$Pos, nt=normalized$nt, swinsor=normalized$swinsor, swinsor.sd=normalized$swinsor.sd) #Keep only relevant columns

    normalized$swinsor[is.nan(normalized$swinsor)] <- NA #Change NaN to NA, for consistency
    normalized$swinsor.sd[is.nan(normalized$swinsor.sd)] <- NA #Change NaN to NA, for consistency

    #If add_to specified, merge with existing normalized data frame:
    if(!missing(add_to)){
    add_to_df <- GR2norm_df(add_to)
    normalized <- merge(add_to_df, normalized, by=c("RNAid", "Pos", "nt"), suffixes=c(".old",".new"))
    }
    ###

    normalized <- normalized[order(normalized$RNAid, normalized$Pos),]
    normalized_GR <- norm_df2GR(normalized)

    normalized_GR
}
